bpo-20082: fix misbehavior of buffered writes to raw files in append mode

embray opened the pull request:
This is a patch for a bug I apparently wrote many years ago but never converted to a PR.

<!-- issue-number: [bpo-20082](https://bugs.python.org/issue20082) -->
https://bugs.python.org/issue20082
<!-- /issue-number -->

embray committed:
bpo-20082: fix misbehavior of buffered writes to raw files in append mode

ZackerySpytz commented on Lib/test/test_io.py:
```diff
@@ -2396,6 +2396,31 @@ def test_interleaved_readline_write(self):
+                f.seek(0)
+
+                # This read previously returned b'tBst testA' but that is
+                # inccorect if the underlying raw file is in append mode;
```
Typo.

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1767,6 +1778,27 @@ _bufferedwriter_reset_buf(buffered *self)
+        }
+        Py_DECREF(mode);
+    } else {
+        PyErr_Clear();
```
Exceptions like `MemoryError` or `KeyboardInterrupt` should not be cleared. `PyErr_ExceptionMatches()` could be used instead of blindly clearing any exception.

embray replied:
The original patch for this was written so long ago, I don't even remember why there's a `PyErr_Clear()` there, but I don't think it's even necessary.

embray replied:
Now I remember why that was there.  Fixed in 08661e3

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1767,6 +1778,27 @@ _bufferedwriter_reset_buf(buffered *self)
+    if (mode != NULL) {
+        /* Raw fileobj has no mode string so as far as we can know it has
+           normal write behavior */
+        if (PyUnicode_FindChar(mode, 'a', 0, PyUnicode_GET_LENGTH(mode), 1) != -1) {
```
Please limit lines to 79 characters (PEP 7).

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1767,6 +1778,27 @@ _bufferedwriter_reset_buf(buffered *self)
+{
+    PyObject *mode;
+
+    mode = _PyObject_GetAttrId(self->raw, &PyId_mode);
```
```suggestion
    PyObject *mode = _PyObject_GetAttrId(self->raw, &PyId_mode);
```

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1767,6 +1778,27 @@ _bufferedwriter_reset_buf(buffered *self)
+           normal write behavior */
+        if (PyUnicode_FindChar(mode, 'a', 0, PyUnicode_GET_LENGTH(mode), 1) != -1) {
+            self->appending = 1;
+        } else {
```
```suggestion
        }
        else {
```
The `}` should go on its own line (according to PEP 7).

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1204,6 +1205,33 @@ buffered_tell(buffered *self, PyObject *Py_UNUSED(ignored))
+
+    if (self->writable) {
+        res = _bufferedwriter_flush_unlocked(self);
+        if (res == NULL)
```
Please add braces (PEP 7).

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1204,6 +1205,33 @@ buffered_tell(buffered *self, PyObject *Py_UNUSED(ignored))
+static PyObject *
+_buffered_seek_unlocked(buffered *self, Py_off_t target, int whence)
+{
+    Py_off_t n;
```
This declaration can be moved down to where the variable is used.

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1795,6 +1827,8 @@ _io_BufferedWriter___init___impl(buffered *self, PyObject *raw,
     self->readable = 0;
     self->writable = 1;
 
+    _bufferedwriter_set_append(self);
```
This call should be checked for failure.

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -2309,6 +2351,8 @@ _io_BufferedRandom___init___impl(buffered *self, PyObject *raw,
     self->readable = 1;
     self->writable = 1;
 
+    _bufferedwriter_set_append(self);
```
This call should be checked for failure.

ZackerySpytz commented on Modules/_io/bufferedio.c:
```diff
@@ -1767,6 +1778,27 @@ _bufferedwriter_reset_buf(buffered *self)
     self->write_end = -1;
 }
 
+static void
```
This function should be checked for failure, so it should not return `void`.

embray replied:
I'm not so sure.  What is its failure state?  It either sets `self->appending = 1` or `self->appending = 0`.  Neither case represents a failure or success.  AFAICT all calls within it check for failure or success and are handled appropriately.

serhiy-storchaka replied:
`getattr()` can raise an arbitrary exception, for example MemoryError or KeyboardInterrupt. They should not be cleared.

BTW, in 3.13 you can use `PyObject_GetOptionalAttr()` which automatically silences an AttributeError. It exists under other name as private API in older versions.

`PyUnicode_FindChar()` also can raise an exception, according to the specification, although it is less likely after getting rid of legacy strings.

embray committed:
trivial fixes addressing review comments

embray committed:
fix I/O error when flushing writes to raw files in append mode

if the raw file is appending there is no reason to attempt to rewind it before
flushing

embray committed:
fix possible unhandled AttributeError (e.g. when the raw file does not have a mode attribute) and ensure that the mode is in fact a string

embray committed:
use PyObject_GetOptionalAttr and fix error handling on PyUnicode_FindChar
